### PR TITLE
Tool 729/bundle id expand fix

### PR DESCRIPTION
**Pull request by BirmacherAkos**
Original PR: https://github.com/bitrise-steplib/steps-ios-auto-provision/pull/77
Moved here as it should be in a package.

---

Update the bundle ID resolve logic, and make it public.

### CODE REVIEW

File: xcodeproj/xcodeproj.go [hunk 1]
```diff
@@ -149,33 +157,73 @@ func resolve(bundleID string, buildSettings serialized.Object) (string, error) {
...
+
+		envKey = envKey[:len(envKey)-1]
+		removedChar++
+	}
+	return strings.Replace(bundleID, envKey, envValue, -1), nil
+
+}
 
-	return prefix + envValue + suffix, nil
+func envInBuildSettings(envKey string, buildSettings serialized.Object) (string, bool) {
```
Comment (BirmacherAkos) [inline xcodeproj/xcodeproj.go]: We need this, because an env could be empty string too: "".

File: xcodeproj/xcodeproj.go [hunk 2]
```diff
@@ -149,33 +155,71 @@ func resolve(bundleID string, buildSettings serialized.Object) (string, error) {
...
 
-	pattern := `(.*)\$\((.*)\)(.*)`
-	re := regexp.MustCompile(pattern)
-	match := re.FindStringSubmatch(bundleID)
-	if len(match) != 4 {
-		return "", fmt.Errorf("%s does not match to pattern: %s", bundleID, pattern)
+// expandComplexEnv expands the env with the "[$][{(][^$]*[)}]" regex
+// **Example:** `prefix.$(ENV_KEY:rfc1034identifier).suffix.$(ENV_KEY:rfc1034identifier)` **=>** `auto_provision.ios-simple-objc.suffix.ios-simple-objc`
+func expandComplexEnv(bundleID string, buildSettings serialized.Object) (string, error) {
+	r, err := regexp.Compile("[$][{(][^$]*[)}]")
```
Comment (godrei) [inline xcodeproj/xcodeproj.go]: [$][{(][^$]*[)}] -> `[$][{(][^$]*?[)}]`

?: Quantifier — Matches between zero and unlimited times, as few times as possible, expanding as needed (lazy)

Given this test string `${test}case}`
The original matches: `${test}case}`
The suggested: `${test}`

File: xcodeproj/xcodeproj.go [hunk 3]
```diff
@@ -149,33 +155,71 @@ func resolve(bundleID string, buildSettings serialized.Object) (string, error) {
...
-		}
 		return "", err
 	}
+	if !r.MatchString(bundleID) {
+		return "", fmt.Errorf("failed to match regex [$][^$]* for %s", bundleID)
+	}
+	envKey := r.FindString(bundleID)
+
+	var envValue string
+	var removedChar int
```
Comment (godrei) [inline xcodeproj/xcodeproj.go]: why do we need this var?

File: xcodeproj/xcodeproj_test.go [hunk 4]
```diff
@@ -11,25 +11,81 @@ import (
 )
 
 func TestResolve(t *testing.T) {
-	t.Log("resolves bundle id in format: prefix.$(ENV_KEY:rfc1034identifier).suffix")
+	t.Log("resolves bundle id in format: prefix.$ENV_KEY")
```
Comment (godrei) [inline xcodeproj/xcodeproj_test.go]: add new test case `${test}case}`

File: xcodeproj/xcodeproj.go (hunk 2 above)
Comment (BirmacherAkos) [reply xcodeproj/xcodeproj.go]: Fixed

File: xcodeproj/xcodeproj.go (hunk 3 above)
Comment (BirmacherAkos) [reply xcodeproj/xcodeproj.go]: We don't. It was left there by mistake.

Comment (BirmacherAkos) [reply xcodeproj/xcodeproj.go]: Fixed

File: xcodeproj/xcodeproj_test.go (hunk 4 above)
Comment (BirmacherAkos) [reply xcodeproj/xcodeproj_test.go]: Fixed

**Pull request merged**